File: examples/py/lasso_path.py
import h2oaiglm as h2oaiglm
import scipy as sp
from numpy import abs, exp, float32, float64, log, max, sum, zeros
from numpy.random import rand, randn


'''
LassoPath

   minimize    (1/2) ||Ax - b||_2^2 + \lambda ||x||_1

   for 50 values of \lambda.
   See <h2oaiglm>/matlab/examples/lasso_path.m for detailed description.
'''

def LassoPath(m, n, gpu=True, double_precision=False, nlambda=50):
  # set solver cpu/gpu according to input args
  if gpu and h2oaiglm.SolverGPU is None:
    print("\nGPU solver unavailable, using CPU solver\n")
    gpu=False

  Solver = h2oaiglm.SolverGPU if gpu else h2oaiglm.SolverCPU

  # random matrix A
  A=randn(m,n)

  # cast A as float/double according to input args
  A=A if double_precision else float32(A)

  # true x vector, ~20% zeros
  x_true=(randn(n)/n)*float64(randn(n)<0.8)

  # b= A*x_true + v (noise)
  b=A.dot(x_true)+0.5*randn(m)

  # lambda_max
  lambda_max = max(abs(A.T.dot(b)))



  # f(Ax) = ||Ax - b||_2^2
  f = h2oaiglm.FunctionVector(m,double_precision=double_precision)
  f.b[:]=b[:]
  f.h[:]=h2oaiglm.FUNCTION["SQUARE"]

  # g(x) = 0.2*lambda_max*||x||_1
  g = h2oaiglm.FunctionVector(n,double_precision=double_precision)
  g.a[:] = 0.2*lambda_max 
  g.h[:] = h2oaiglm.FUNCTION["ABS"]


  # store results for comparison 
  x_prev = zeros(n)

  # timer
  runtime = 0.

  # use problem data A to create solver

  #s = Solver(sp.sparse.csr_matrix(A))
  #s = Solver(sp.sparse.csc_matrix(A))
  s = Solver(A)

  for i in range(nlambda):
    _lambda= exp( (log(lambda_max)*(nlambda-1-i)+1e-2*log(lambda_max)*i )/ (nlambda-1))

    g.c[:]=_lambda

    # solve
    s.solve(f, g)

    # add run time
    runtime += s.info.solvetime

    # copy
    x_curr=s.solution.x

    # check stopping condition
    if max(abs(x_prev-x_curr)) < 1e-3* sum(abs(x_curr)):
      break
    
    x_prev[:]=x_curr[:]


  # tear down solver in C++/CUDA
  s.finish()

  return runtime

if __name__ == "__main__":
   print("Solve time:\t{:.2e} seconds".format(LassoPath(200000,1000)))

File: examples/cpp/lasso_path.cpp
#include <stdio.h>
#include <stdlib.h>

#include <limits>
#include <random>
#include <vector>

#include "matrix/matrix_dense.h"
#include "pogs.h"
#include "timer.h"

using namespace pogs;

template <typename T>
T MaxDiff(std::vector<T> *v1, std::vector<T> *v2) {
  T max_diff = 0;
#ifdef _OPENMP
#pragma omp parallel for reduction(max : max_diff)
#endif
  for (unsigned int i = 0; i < v1->size(); ++i)
    max_diff = std::max(max_diff, std::abs((*v1)[i] - (*v2)[i]));
  return max_diff;
}

template <typename T>
T Asum(std::vector<T> *v) {
  T asum = 0;
#ifdef _OPENMP
#pragma omp parallel for reduction(+ : asum)
#endif
  for (unsigned int i = 0; i < v->size(); ++i)
    asum += std::abs((*v)[i]);
  return asum;
}

// LassoPath
//   minimize    (1/2) ||Ax - b||_2^2 + \lambda ||x||_1
//
// for 100 values of \lambda.
// See <pogs>/matlab/examples/lasso_path.m for detailed description.
template <typename T>
double LassoPath(size_t m, size_t n) {
  unsigned int nlambda = 100;
  std::vector<T> A(m * n);
  std::vector<T> b(m);
  std::vector<T> x_last(n, std::numeric_limits<T>::max());

  // Generate data
  fprintf(stdout,"BEGIN FILL DATA\n");
  double t0 = timer<double>();
  std::default_random_engine generator;
  std::uniform_real_distribution<T> u_dist(static_cast<T>(0),
                                           static_cast<T>(1));
  std::normal_distribution<T> n_dist(static_cast<T>(0),
                                     static_cast<T>(1));

  for (unsigned int i = 0; i < m * n; ++i)
    A[i] = n_dist(generator);

  std::vector<T> x_true(n);
  for (unsigned int i = 0; i < n; ++i)
    x_true[i] = u_dist(generator) < 0.8 ? 0 : n_dist(generator) / n;

  for (unsigned int i = 0; i < m; ++i) {
    for (unsigned int j = 0; j < n; ++j)
      b[i] += A[i * n + j] * x_true[j];
      // b[i] += A[i + j * m] * x_true[j];
    b[i] += static_cast<T>(0.5) * n_dist(generator);
  }

  T lambda_max = static_cast<T>(0);
  for (unsigned int j = 0; j < n; ++j) {
    T u = 0;
    for (unsigned int i = 0; i < m; ++i)
      //u += A[i * n + j] * b[i];
      u += A[i + j * m] * b[i];
    lambda_max = std::max(lambda_max, std::abs(u));
  }

  // Set up pogs datastructure.
  pogs::MatrixDense<T> A_('r', m, n, A.data());
  pogs::PogsDirect<T, pogs::MatrixDense<T> > pogs_data(A_);
  std::vector<FunctionObj<T> > f;
  std::vector<FunctionObj<T> > g;
  f.reserve(m);
  g.reserve(n);

  for (unsigned int i = 0; i < m; ++i)
    f.emplace_back(kSquare, static_cast<T>(1), b[i]);

  for (unsigned int i = 0; i < n; ++i)
    g.emplace_back(kAbs);

  double t1 = timer<double>();
  fprintf(stdout,"END FILL DATA\n");

  fprintf(stdout,"BEGIN SOLVE\n");
  double t = timer<double>();
  for (unsigned int i = 0; i < nlambda; ++i) {
    T lambda = std::exp((std::log(lambda_max) * (nlambda - 1 - i) +
        static_cast<T>(1e-2) * std::log(lambda_max) * i) / (nlambda - 1));

    for (unsigned int i = 0; i < n; ++i)
      g[i].c = lambda;

    pogs_data.Solve(f, g);

    std::vector<T> x(n);
    for (unsigned int i = 0; i < n; ++i)
      x[i] = pogs_data.GetX()[i];

    if (MaxDiff(&x, &x_last) < 1e-3 * Asum(&x))
      break;
    x_last = x;
  }
  double tf = timer<double>();
  fprintf(stdout,"END SOLVE: type 1 tfd %g ts %g\n",t1-t0,tf-t);

  return tf-t;
}

template double LassoPath<double>(size_t m, size_t n);
template double LassoPath<float>(size_t m, size_t n);

